### PR TITLE
added container services

### PR DESCRIPTION
**Pull request by bentheiii**

### CODE REVIEW

**Comment by bharel** [comment]
How about this?

```
@property
@abstractmethod
def containers(self):
  """Return all containers by initialization order"""
```

File: yellowbox/subclasses.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    def stop(self, signal='SIGTERM'):
+        for c in reversed(self.containers):
+            if not is_alive(c):
+                continue
+            c.kill(signal)
+            c.wait(timeout=_DEFAULT_TIMEOUT)
+            c.reload()
+            if self._remove:
+                c.remove()
```
Comment (bharel) [inline yellowbox/subclasses.py:39]: volume=True

File: yellowbox/subclasses.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    def disconnect(self, network: Network, **kwargs):
+        for ec in reversed(self._endpoint_containers):
+            network.disconnect(ec, **kwargs)
+            ec.reload()
+
+
+class SingleEndpointService(ContainerService):
+    @property
+    @abstractmethod
+    def _single_endpoint(self) -> Container:
```
Comment (bharel) [inline yellowbox/subclasses.py:63]: "_endpoint"?

File: yellowbox/subclasses.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        pass
+
+    def _endpoint_containers(self) -> Sequence[Container]:
```
Comment (bharel) [inline yellowbox/subclasses.py:66]: `Tuple[Container]`

File: yellowbox/containers.py [hunk 4]
```diff
@@ -108,10 +108,14 @@ def killing(container: _CT, *, timeout: float = _DEFAULT_TIMEOUT,
...
 def create_and_pull(docker_client: DockerClient, image, command=None, **kwargs) -> Container:
     try:
-        return docker_client.containers.create(image=image, command=command, **kwargs)
+        ret = docker_client.containers.create(image=image, command=command, **kwargs)
     except ImageNotFound:
         docker_client.images.pull(image, platform=None)
-        return docker_client.containers.create(image=image, command=command, **kwargs)
+        ret = docker_client.containers.create(image=image, command=command, **kwargs)
+
+    if isinstance(ret, Container):
```
Comment (bharel) [inline yellowbox/containers.py:116]: When is container not going to be a container?

File: yellowbox/subclasses.py (hunk 1 above)
Comment (bentheiii) [reply yellowbox/subclasses.py:39]: is in a separate issue

File: yellowbox/subclasses.py (hunk 3 above)
Comment (bentheiii) [reply yellowbox/subclasses.py:66]: no need to restrict, Sequence is enough

File: yellowbox/containers.py (hunk 4 above)
Comment (bentheiii) [reply yellowbox/containers.py:116]: if detach=False

Comment (bharel) [reply yellowbox/containers.py:116]: Stupid api >.<

**Review by bharel — APPROVED** [review]

**Pull request merged**